[mlir][Properties] Shorten "Property" to "Prop" in most places

krzysz00 opened the pull request:
Since the property system isn't currently in heavy use, it's probably the right time to fix a choice I made when expanding ODS property support.

Specifically, most of the property subclasses, like OptionalProperty or IntProperty, wrote out the word "Property" in full. The corresponding classes in the Attribute hierarchy uses the short-form "Attr" in those cases, as in OptionalAttr or DefaultValuedAttr.

This commit changes all those uses of "Property" to "Prop" in order to prevent excessively verbose tablegen files that needlessly repeat the full name of a core concept that can be abbreviated.

So, this commit renames all the FooProperty classes to FooProp, and keeps the existing names as alias with a Deprecated<> on them to warn people.

In addition, this commit updates the documentation around properties to mention the constraint support.

krzysz00 commented:
So as to note upcoming plans:
- EnumProp probably wants to move to `EnumAttr` and grow a real predicate.
- `IndexProp` and `APIntProp` (name subject to debate) should exist
- Once the infrastructure feels like it's in a solid shape, I intend to go on a rampage through the upstream dialects where things that could reasonably be properties (ex. integer scalars) get turned into properties

zero9178 commented on mlir/docs/DefiningDialects/Operations.md:
```diff
@@ -102,8 +102,9 @@ their semantics via a special [TableGen backend][TableGenBackend]:
-    properties that are inherent to operations. This will be expanded to a
-    `PropertyConstraint` class or something similar in the future.
+    properties that are inherent to operations. These properties can have
+    constraints imposed them using their `predicate` field or the
```
```suggestion
    constraints imposed on them using the `predicate` field or the
```

zero9178 commented on mlir/tools/mlir-tblgen/OpFormatGen.cpp:
```diff
@@ -113,13 +113,13 @@ struct AttributeLikeVariable : public VariableElement {
     if (const auto *prop = dyn_cast<PropertyVariable>(this)) {
       return prop->getVar()->prop.getBaseProperty().getPropertyDefName() ==
-             "UnitProperty";
+             "UnitProp";
```
Isn't this going to fail now for `UnitProperty`?

krzysz00 replied:
Good point, fixed

zero9178 reviewed (commented):
I am personally in favour of this and besides the past, there probably isn't a better time to do this change than now. I also really like that this only deprecates the old names such that downstream users are notified on their next bump and it can then be removed later.

Could we move the documentation changes that are specific to predicates (not renaming) to either the other PR or a new PR/commit? I don't think they should be intermingled here.

Mogball reviewed (approved):
Makes sense

krzysz00 committed:
[mlir][Properties] Shorten "Property" to "Prop" in most places

Since the property system isn't currently in heavy use, it's probably
the right time to fix a choice I made when expanding ODS property
support.

Specifically, most of the property subclasses, like OptionalProperty
or IntProperty, wrote out the word "Property" in full. The
corresponding classes in the Attribute hierarchy uses the short-form
"Attr" in those cases, as in OptionalAttr or DefaultValuedAttr.

This commit changes all those uses of "Property" to "Prop" in order to
prevent excessively verbose tablegen files that needlessly repeat the
full name of a core concept that can be abbreviated.

So, this commit renames all the FooProperty classes to FooProp, and
keeps the existing names as alias with a Deprecated<> on them to warn
people.
